internal/services/v1/relationships.go

```diff
@@ -162,31 +161,23 @@ func (ps *permissionServer) WriteRelationships(ctx context.Context, req *v1.Writ
 
 	// Check for duplicate updates and create the set of caveat names to load.
 	updateRelationshipSet := util.NewSet[string]()
-	referencedCaveatNamesWithContext := util.NewSet[string]()
 	for _, update := range req.Updates {
 		tupleStr := tuple.StringRelationship(update.Relationship)
 		if !updateRelationshipSet.Add(tupleStr) {
 			return nil, rewriteError(
 				ctx,
-				status.Errorf(
-					codes.InvalidArgument,
-					"found duplicate update operation for relationship %s",
-					tupleStr,
-				),
+				NewDuplicateRelationshipErr(update),
 			)
 		}
 
-		// Only load the caveat if we need its type information for context checking.
-		if hasNonEmptyCaveatContext(update) {
-			if !ps.caveatsEnabled {
+		if !ps.caveatsEnabled {
+			if update.Relationship.OptionalCaveat != nil && update.Relationship.OptionalCaveat.CaveatName != "" {
```

Comment:
the previous logic seemed incorrect, I think? It checked if caveat was non-empty first instead of checking that the functionality was disabled first. The `hasNonEmptyCaveatContext ` method checked relationships had caveat *and context was not empty*, which means it would have allowed writing empty rels with empty context, despite the disablement flag?

The new logic seems correct 👍🏻

Reply:
As discussed offline, the caveat itself would still have been checked below, but this was deemed safer too :)

Reply:
👍🏻